src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = state;
+
+  actions.forEach((action) => {
+    let nextState;
+
+    switch (action.type) {
+      case 'clear':
+        nextState = {};
+        break;
+      case 'addProperties':
+        nextState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        nextState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          if (nextState.hasOwnProperty(key)) {
+            delete nextState[key];
+          }
+        });
+        break;
+      default:
+        nextState = { ...currentState };
+        break;
```

Comment:
Consider using a 'switch' statement for a more readable and organized structure when handling different action types. Additionally, a 'default' case should be included for error handling.